### ISSUE TITLE
Port to TypeScript

### ISSUE THREAD

**Issue by nickmccurdy** [reduxjs/redux#3500]
**Do you want to request a _feature_ or report a _bug_?**

No

**What is the current behavior?**

Redux publishes a manually created index.d.ts file. This also affects reselect, redux-devtools, and redux-thunk.

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem via https://jsfiddle.net or similar.**

Explanation of why this isn't recommended: https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html

To summarize, type declarations should either be generated from TypeScript source code or shared in DefinitelyTyped, which allows contributors to edit the types without having to edit the original package. This can also help alleviate maintenance and versioning issues regarding Redux users that write their code in TypeScript. Many issues about TypeScript compatibility could be moved to the dedicated DefinitelyTyped community if the types were separated, and the types could have breaking changes released without interfering with non-TypeScript users.

**What is the expected behavior?**

Redux either removes its type declaration (should be a breaking change) and moves them to DefinitelyTyped (recommended) or is ported to TypeScript (could be difficult and requires additional maintenance, but makes the types more accurate).

**Which versions of Redux, and which browser and OS are affected by this issue? Did this work in previous versions of Redux?**

3.4.0+ (all versions that publish index.d.ts)

**Comment by jamestharpe** [comment]
I'd like to see redux ported to TypeScript because TypeScript is the only way to ensure types are always in sync.

Types that are always in sync prevent issues from being reported, improve the overall developer experience, and increase the overall potential for adoption of the library.

The general benefits of types are that they improve the quality and productivity of the developer experience, even for non TypeScript developers. Types help catch bugs, enable better tooling, and enhance understanding. 

Types provide these benefits most effectively when they perfectly match the codebase. Leaving types to a separate community all but ensures a lag between library changes and the related type definition changes, making it an overall less effective approach to providing types and therefore lessening the overall value of the library.

Regarding effort, because TypeScript is a superset of JavaScript the code base can be converted gradually. Gradual conversion means ongoing efforts are only minimally disrupted. Converting to TypeScript can also be done as a nonbreaking change, meaning it is also less effort for developers to upgrade to.

Finally, TypeScript it's self has a vibrant community. Conversion to TypeScript may garner additional contributions to help with the effort.

**Comment by markerikson** [comment]
Relevant Twitter discussion thread:

https://twitter.com/acemarke/status/1160243663370371073

**Comment by markerikson** [comment]
@jamestharpe : realistically, Redux is not going to be ported to TS any time soon.

Tim and I don't have enough TS experience to do the porting or maintenance ourselves, and even if someone else does the work, that basically makes the code impossible for us to maintain.  That's already happened over in Redux Starter Kit - it was ported by several users, and now I am effectively unable to make meaningful changes to the code myself.

**Comment by janhesters** [comment]
@markerikson Redux might be small enough to learn TypeScript 😊

That being said, I prefer types bundled with the package instead of DefinitelyTyped. The reason is simple. It's a sync issue. If the types are bundled with the package, they are usually updated with breaking changes and additions as opposed to after. I've helped moving the React Navigation types into the package for exactly that reason.

**Comment by ricokahler** [comment]
Coming from a typescript fanboy, I don’t think redux should be ported to typescript. Especially if the maintainers don’t have a strong TS background.

I personally think the typing should be removed from this library and moved to definitely typed. I think the community will do a great job maintaining them.

Just look at react itself. Definitely typed manages those and they’re great. Given the popularity of this library, I don’t think there will be much lag time between versions even.

**Comment by timdorr** [comment]
I've got a pretty handle on our types at this point. There are some things I would like to add (more inferred types, for instance), but those are nice-to-haves, not things that make the types unusable or inaccurate. I do get the concerns around independent versioning, but I don't think there's any particular need to update them right now, so that's not really a practical concern.

I've actually already tried this with redux-thunk and it's a complete rats nest. The linting alone was a giant hassle; other dependent packages hadn't been updated in a while and were out of compliance with the latest config. And [there are a *lot* of things that rely on redux](https://github.com/DefinitelyTyped/DefinitelyTyped/search?l=JSON&q=dependencies+redux) that are going to trip those same set of issues. It's a giant mess to unravel.

**Comment by nickmccurdy** [comment]
Overall I think moving to DefinitelyTyped is the best option. Yes, TS generates more accurate types, but it’s too much work for Redux to migrate without a rewrite. We can improve the current state of typings by maintaining them separately. Bundling like we’re doing now is not recommended by Microsoft.

**Comment by markerikson** [comment]
@janhesters : I _have_ learned TS.

I just don't know TS well enough to be able to do stuff like this:

https://github.com/reduxjs/redux/blob/63dda816402d11766e76f92eb180cfeaba84b4c0/index.d.ts#L375-L384

or this, from Redux Starter Kit:

```ts
type CaseReducerActions<CR extends SliceCaseReducers<any, any>> = {
  [T in keyof CR]: CR[T] extends (state: any) => any
    ? PayloadActionCreator<void>
    : (CR[T] extends (state: any, action: PayloadAction<infer P>) => any
        ? PayloadActionCreator<P>
        : CR[T] extends { prepare: PrepareAction<infer P> }
        ? PayloadActionCreator<P, string, CR[T]['prepare']>
        : PayloadActionCreator<void>)
}

type NoInfer<T> = [T][T extends any ? 0 : never];
type SliceCaseReducersCheck<S, ACR> = {
    [P in keyof ACR] : ACR[P] extends {
        reducer(s:S, action?: { payload: infer O }): any 
    } ? {
        prepare(...a:never[]): { payload: O }
    } : {

    }
}
```

**Comment by nickmccurdy** [comment]
Screw it I’m porting everything to Haskell 😂

**Comment by ee0pdt** [comment]
For libraries that are not written in TS I’ve noticed a worrying lag in types being added to definitely typed vs maintained in the repo. I’ve also seen more bugs and errors in the typings which lead to issues getting raised in the library even though these only relate to the types. 

Based on the latest JS surveys and extrapolating out I would expect that this year around 55% of JS devs will have used TS and be happy to use it again. At the point that the majority of JS devs are actually TS devs, every JS library needs to consider TS as a majority use-case. This means eventual death to any lib that does not have good typings available.

But, I do appreciate that types for non-trivial libraries are a massive undertaking so I do see why the maintainers would want to discuss this.

**Comment by oriSomething** [comment]
If the maintainers aren't TS users, I think maintaining types as part of release is problematic because:
1. It's pain releasing patch versions only because in types mistakes, I saw it a lot in `immer` which sometimes there where 3 releases in a row only for types fixes
2. As someone who contribute to DT, I think it's less painful than contribute to projects themselves especially when the maintainer isn't a heavy TS user
3. For popular projects, mostly DT definitions are better than project's original when it's not TS. For example, React written in Flow, but as someone who uses both Flow and TS, mostly TS definitions that created by community are superior

**Comment by jamestharpe** [comment]
Given the maintainer’s lack of interest in porting to TypeScript, I think DT is the way to go.

* Lag should be roughly the same 
* Redux package size will be slightly smaller 
* Fewer issues/concerns for Redux maintainers will help focus on new features

**Comment by nickmccurdy** [comment]
It seems like moving to DefinitelyTyped is preferred by most users and maintainers.

@acemarke @timdorr How would you like to proceed? I’m thinking we could create a branch/PR that removes Redux types, then PR the Redux package types to DefinitelyTyped (may involve some extra code style and test work to meet their review criteria). Once that’s published, it should be safe to merge the removal of the declaration file as a major release (since it’s a breaking change for TypeScript users).

**Comment by markerikson** [comment]
FWIW, Daniel Rosenwasser from the TS team says they should be moved to DT:

https://twitter.com/drosenwasser/status/1160628533166628864

I personally would be in favor of moving them to DT, but I know @timdorr has expressed interest in keeping them here.  So, I'll leave it up to him to decide.

**Comment by markerikson** [comment]
All right, so I'll throw this out there _just_ to bring it up:

@timdorr , do we ever foresee the rest of the Redux libs getting rewritten in TS?

I would assume no, but just want to get your thoughts on that for the record.

**Comment by nickmccurdy** [comment]
I think the benefits of maintaining code and types together would outweigh the cost of migrating to TypeScript for smaller packages like redux-thunk and reselect (which has extremely complex types that are harder to test because they're separate from the source code). It still might backfire if the maintainers don't have TypeScript experience, but they're simple enough that they shouldn't go very out of sync.

**Comment by timdorr** [comment]
My concern with switching anything to TS is approachability. Even after transpilation and bundling, [our code is still quite readable](https://unpkg.com/browse/redux@4.0.4/lib/redux.js). And given the size of the library and its importance in the ecosystem, that's something I do not ever want to compromise on. 

In addition, our types are also pretty well documented and quite functional at this point. Again, we really only have improvements to make to them, not fixes. So other than the decorum of moving them elsewhere, I don't see that changing how much they're developed or often they're released. 

For redux at least, I'm pretty much fully against this. It makes a lot more sense for things like redux-thunk (which I had huge problems when I last tried it...), but not for this small, central library.

**Comment by ricokahler** [comment]
> In addition, our types are also pretty well documented and quite functional at this point. Again, we really only have improvements to make to them, not fixes. So other than the decorum of moving them elsewhere, I don't see that changing how much they're developed or often they're released.

For clarification on [my above comment](https://github.com/reduxjs/redux/issues/3500#issuecomment-520177326), if there is no real problem, then I wouldn't fix it.

**Comment by timdorr** [comment]
![](https://media.giphy.com/media/jL0c2TrfDUz9S/giphy.gif)

**Comment by nickmccurdy** [comment]
@timdorr Understood. Are you still comfortable with moving the types to DefinitelyTyped though? Work has already started in https://github.com/reduxjs/redux/pull/3501.

**Comment by timdorr** [comment]
Just making a copy of them? Sure, but I'm not sure if that would create problems for the other type packages that depend on them.

**Comment by nickmccurdy** [comment]
No, I mean deleting them from Redux as a breaking change and major release.

**Comment by timdorr** [comment]
No, I said I'm fully against that. I want to keep them here.

**Comment by nickmccurdy** [comment]
Can we discuss why? This is not recommended by the TypeScript documentation and I think it’s been causing versioning and maintenance issues with Redux. Removing types is the simplest way to solve the problem without porting the code to TypeScript (which I don’t think is a good idea in this situation).

**Comment by gengjiawen** [comment]
Keep typing here will save us version problem and easy to change.

**Comment by timdorr** [comment]
I've already stated my reasons above.

Honestly, DT is just a maze of tooling full of landmines and gotchas. It's a half-gig of git data with 6000+ types and nearly 3000 open issues. It's extremely hard to get into, and the tooling that's in place doesn't make it particularly easy to get started. When I tried to move in the redux-thunk types (one project where I _do_ support the move), I couldn't figure it out. All the docs were unhelpful and the tooling gave no reasoning for why unrelated types were failing. 

Again, one thing I really like about Redux and is very important to its community, is its approachability as a code base. DT is the opposite of this, IMO. Don't get me wrong, I think it's great to have a central clearinghouse for type definitions for projects unaware of TS. But the nature of the Redux community runs counter to that. 

I think the perception of TS being a maintenance issue might be due to the past few years where I didn't know any TS and relied heavily on the community. I'm at the level where I may not be the best TS coder, but I think I've gotten pretty good. Certainly good enough where I can offer more meaningful review of type PRs. 

For all intents and purposes, Redux is "done" right now. We don't have plans for another major (including the types), so the types can be the driver on moving the version numbers forward. That's totally fine in my mind. It doesn't create a maintenance problem because the actual code we distribute stays the same and we can easily check that it's not changing from release to release (I always run `npm pack` before publishing a release). 

One thing that might be an interesting experiment would be a parallel implementation of the code in TypeScript. Basically, move the types out of the definition files to the code itself, side-by-side with the plain JS implementation. That's definitely a more complex setup, but it may make the type def generation easier and less like the fancy examples that Mark posted above. Again, this is made feasible by the code being "done" and unlikely to receive major changes in the future. 

Either that or we do actually rewrite into TS and have some pre-commit hooks to transpile down to plain JS and store that in the repo. 

Let me do a Twitter poll for science.

**Comment by timdorr** [comment]
https://twitter.com/timdorr/status/1161278569038974977

**Comment by nickmccurdy** [comment]
TypeScript is still designed to keep separate types in DefinitelyTyped, it's still a better option even if it's more difficult to use. DefinitelyTyped will get easier to contribute to over time as it matures, and it already has some very useful tooling that makes this kind of type work easier.

Also, I don't understand why there would be a second implementation in TypeScript, unless it had breaking changes to the API. If we can manage to rewrite Redux with full API compatibility (and our tests should make that easy to determine), we can just replace the JS implementation with the TS one, and it will still be fully usable by JS users (TypeScript packages are actually published as plain JS, with the TS declaration files automatically generated, so they work well with both languages).

**Comment by markerikson** [comment]
I would be interested in seeing what a straight TS conversion looks like, even if only for point of comparison.

Ideally, it would start by renaming one file at a time from `.js` to `.ts` to maintain the existing change history, then copy and apply the relevant types from our typings.

Anyone wanna tackle that and throw it up as a PR?

**Comment by timdorr** [comment]
@nickmccurdy I'm not sure about reselect and redux-devtools, as I don't maintain those projects. With redux-thunk, the plan is to move them out with a 3.0 release: reduxjs/redux-thunk#180

**Comment by nickmccurdy** [comment]
@markerikson By porting to TypeScript, we will have to fix type errors for the internal code (such as functions that aren't exposed in the API and variables within functions) and make the type declarations make the code exactly (right now the type tests are the only thing keeping them in sync). This involves more effort than just working on type declarations and may take a while, but it's worth it for the consistency between source code and types that we'll have in the long run. Also note that how safe the types are and how much effort it takes depends on our TypeScript config (I generally recommend starting with `{ "compilerOptions": { "strict": true} }`).

@timdorr What's the plan for redux-thunk and does it need help? I'll also make issues for reselect and redux-devtools.

**Comment by timdorr** [comment]
I need to try it again. When I tried it last, it tripped linting errors on completely unrelated packages (ones that didn't use redux-thunk types at all!). I was also very new to TS at the time, so I wasn't equipped to debug. I just need to re-attempt at this point.

**Comment by nickmccurdy** [comment]
Since my original suggestions aren't happening, I'll close this. Feel free to continue discussion here or create related issues.

**Comment by markerikson** [comment]
Well, we actually seem to be discussing the "port to TS" option, which you did originally mention, so I'll reopen it on those grounds.

**Comment by timdorr** [comment]
Yeah, it's roughly a 2:1 split in favor of TS, which I find interesting. I am a bit worried about the 3rd of people that want to keep it in JS, but we should probably get some color on why they chose that option.

**Comment by nickmccurdy** [comment]
Those people can still use Redux in JS projects in the same way (assuming we can figure out how to implement TypeScript without changing the API). It will affect contributors though.

**Comment by markerikson** [comment]
Tweeted a follow-up per Tim's question:

https://twitter.com/acemarke/status/1161453274572165120

**Comment by cellog** [comment]
I have interest in helping here. My first thought is based on https://github.com/reduxjs/redux/issues/3500#issuecomment-520185136

The primary issue I see with these types is there are 0 explanatory comments. I don't know who decided this was a good idea in types, but it's certainly a questionable idea. Just because the compiler understands it doesn't make it easy to understand in case they need to change (as I'm sure you know viscerally Mark :)

Also, based on my experience porting js to ts, I would highly recommend starting with the tests first. I've usually gone the other way, but that was with a project that had no typing at all. We have types here. Once tests are ported to ts, that makes it far simpler to verify correctness when porting the js source to typescript.

In any case, probably step 1 is to modify the tooling to support TS, is there a feature branch to PR against yet?

**Comment by markerikson** [comment]
@cellog : no branches yet that I know of.  Might as well throw one together.

**Comment by nickmccurdy** [comment]
I've started configuring it but haven't made progress on merging the types yet.

**Comment by nickmccurdy** [comment]
@cellog I tried typechecking the tests, they seem to have much more dynamic code and have more type errors than the internals of Redux. I think it would be simpler to get the internal code to typecheck with `"strict": true` (the only useful compiler option we'd want to use from the existing typings/tests), then integrate the type declarations into it, then fix external usage like tests, examples, and the website.

**Comment by cellog** [comment]
got all the tests ported over, but the last one has a pesky type conflict. It turns out `Observable` is not quite compatible with redux's smaller `Observable` type. I will have to look further to find a solution, then we'll be ready for the big switch of the actual source.

**Comment by nickmccurdy** [comment]
Do you want to start a branch together so we don’t have to duplicate work? I started merging the type declarations into the source, but haven’t touched the tests yet.

**Comment by cellog** [comment]
Nick, check out the `ts-conversion` branch, all of the tests have been converted. There is a configuration issue, however. Rollup does not recognize `.ts` files imported from it. Haven't debugged that yet, would be great if you want to tackle that!

**Comment by cellog** [comment]
@nickmccurdy I have some time now, I'll tackle configuration. let me know if you want to develop a plan to split things up

**Comment by nickmccurdy** [comment]
Here's my config so far, but it's extremely simple: https://github.com/nickmccurdy/redux/tree/typescript

I was planning on adding `"strict": true` after merging the declarations in and fixing the initial type errors.

Would you like me to try to move these changes to your branch or keep them separate for now?

**Comment by cellog** [comment]
I've been basing my PRs off of `reduxjs/redux` branch `ts-conversion` to make things simple. I just fixed the config issue for rollup, going to also tackle the eslint config and PR that up in a minute

I'll get back to you shortly about the path forward (like 2 minutes)

**Comment by cellog** [comment]
so #3514 is up, fixes ts, you'll definitely want that to continue. Here's the path I would recommend, for easiest review:

1. merge in the declarations file by file (each in separate PR). `strict: true` is already in the `tsconfig.json` for this branch.

That's it :).

Do you want to split up the files? At this point, it should be pretty straightforward to do the merge

**Comment by cellog** [comment]
quick question: in terms of testing the types themselves, now that we will be using typescript directly, perhaps we could move the typescript files into a single test that simple includes them? It will refuse to compile and break if we change them.

**Comment by nickmccurdy** [comment]
Are you suggesting we keep type error expectations in the type tests, and import the type tests in the Jest tests? I don't think that would work, because you need special tools to check type error expectations. I'd recommend dtslint, which is what Microsoft and DefinitelyTyped used. Alternatively we could keep using typings-tester.

**Comment by cellog** [comment]
ok. The reason I'm asking is I'm not certain that typings-tester is actually working. when I load `test/typescript/store.ts` into VS Code, there are several type errors, but the tests pass

**Comment by nickmccurdy** [comment]
Personally I think it will be useful to keep the type tests around, since they can test types that aren't valid, while letting us safely type our implementation tests using TypeScript

**Comment by cellog** [comment]
ok, so major problem:

in `createStore.ts`, the value returned from `createStore` needs to define a property `[$$observable]()` to be an observer. However, our typescript type requires that the property name be `[Symbol.observable]`. Typescript complains that the field is not set when we use `$$observable`, but if I change it to `Symbol.observable`, then at run-time, the tests fail.

```
 FAIL  test/createStore.spec.ts
  ● createStore › Symbol.observable interop point › should pass an integration test with a common library (RxJS)

    TypeError: You provided an invalid object where a stream was expected. You can provide an Observable, Promise, Array, or Iterable.

      740 |         combineReducers({ foo, bar })
      741 |       )
    > 742 |       const observable = from(store)
          |                          ^
      743 |       const results = []
      744 | 
      745 |       const sub = observable

      at Object.<anonymous>.exports.subscribeTo (node_modules/rxjs/src/internal/util/subscribeTo.ts:27:11)
      at from (node_modules/rxjs/src/internal/observable/from.ts:114:30)
      at Object.<anonymous> (test/createStore.spec.ts:742:26)
```

The only way I see around this is to make the observable field optional in the type definition, which is not great. Any ideas?

**Comment by MichaelDeBoey** [comment]
@markerikson @timdorr CTL moved their typings to DT, but added them as a `dependency` to their package. Could be an interesting move for `reselect`/`redux-thunk`/... too if you want to keep them in JS.
https://blog.johnnyreilly.com/2019/08/symbiotic-definitely-typed.html

**Comment by cellog** [comment]
I've reached a dead end with the `createStore`/`observable` issue. Until that is resolved, the entire conversion is blocked.

**Comment by cellog** [comment]
figured out a great hack to solve this:

```ts
const store: Store<S, A> = ({
    dispatch: dispatch as Dispatch<A>,
    subscribe,
    getState,
    replaceReducer,
    [$$observable]: observable
  } as unknown) as Store<S, A>
```

This is not a good long-term fix, but at least unblocks for now

**Comment by cellog** [comment]
another snag: This babel bug prevents the current implementation of `bindActionCreator` from being typeable: https://github.com/babel/babel/issues/10138

```ts
function bindActionCreator<A extends AnyAction = AnyAction>(
  actionCreator: ActionCreator<A>,
  dispatch: Dispatch
) {
  return function(this: any, ...args: any[]) {
    return dispatch(actionCreator.apply(this, args))
  }
}
```

transforms to:

```js
function bindActionCreator(actionCreator, dispatch) {
  return function () {
    for (var _len = arguments.length, args = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
      args[_key - 1] = arguments[_key];
    }

    return dispatch(actionCreator.apply(this, args));
  };
}
```

I'll see if there is some other way to get the same functionality

**Comment by nickmccurdy** [comment]
Do you think it would be easier for us to switch to compiling with TypeScript instead of Babel? To do that, we could merge #3514 and remove Babel (or if we need syntax features that aren't supported in TypeScript, keep Babel, but remove `@babel/preset-typescript` so TypeScript can sitll handle the types).

**Comment by timdorr** [comment]
I'm really not a fan of using tsc for transpiling. Babel is way ahead of it in terms of correctness and featureset. Plus, preset-env is a lifesaver.

**Comment by cellog** [comment]
I agree. Better is to fix the bug in babel. I'm looking into it now, it looks like they're missing a `replaceWith` on the function after deleting the `this` parameter or `remove` on the `this` parameter. I'm unfamiliar with Babel plugins, so I'm trying to figure out which one is best so I can make a PR

**Comment by cellog** [comment]
found a workaround!

```ts
function bindActionCreator<A extends AnyAction = AnyAction>(
  actionCreator: ActionCreator<A>,
  dispatch: Dispatch
) {
  return function(this: any) {
    return dispatch(actionCreator.apply(this, (arguments as unknown) as any[]))
  }
}
```

**Comment by nickmccurdy** [comment]
Excellent! Would you mind pushing branch(es) up with these type fixes/migrations when you have time? I still have some in progress type fixes, but it seems like it would be easier to adapt them to your code so I can fix types in other APIs.

**Comment by cellog** [comment]
So, just a brief explainer of what's going on with all the PRs above:

I dove down a very deep rabbit hole to solve #3482, and in the process, learned several simple ways to fix a few of the uses of `any` in #3520. So, to make things simple, I have sliced and diced it into 1 PR per converted file. The most complex (unsurprisingly) is `createStore.ts`.

Once the fix for #3482 is merged (#3524) then #3533 will need to be rewritten slightly to support the fixes to `replaceReducer`, but this is a pretty small change.

Next steps after the above slew of PRs are merged:

1. re-open #3535 and convert the configuration as well
2. Once that is merged, file by file, extract the types from `index.d.ts` and put them in the files.
3. Change the `tsconfig.json` to generate `index.d.ts`
4. done with the initial pass of converting to typescript.

At this stage, any BC-breaking type changes could/should be considered prior to pulling the trigger on a release.

Any thoughts on this plan, @timdorr and @markerikson and @nickmccurdy ?

**Comment by markerikson** [comment]
LGTM.

**Comment by MichaelDeBoey** [comment]
I think we should avoid using `any` and try to use `unknown` instead (where possible).

> This is the main value proposition of the `unknown` type: TypeScript won't let us perform arbitrary operations on values of type `unknown`. Instead, we have to perform some sort of type checking first to narrow the type of the value we're working with.

https://mariusschulz.com/blog/the-unknown-type-in-typescript

Happy to make a PR if wanted. 🙂

**Comment by cellog** [comment]
@MichaelDeBoey by all means, yes.

There are a few cases where `any` is used in order to allow for the huge range of possibilities (state values, for instance), but some of these can be tightened up. For instance, I don't think it would be too restrictive to require state to be `object | string | number`.

One example where `any` is probably the right choice: when writing a store enhancer that just passes `preloadedState` unmodified to the underlying store, `any` seems appropriate, as it doesn't interfere with the typing of `createStore` or the return value typing of the store enhancer. What do you think?

I'm not sure `unknown` is what we want for cases like the definition of an action. This would force users to type their actions strictly, and in many cases where people are moving from javascript to typescript, this will require huge PRs right off the bat with a bunch of boilerplate. That would have stopped the last company I worked at from even considering the move. Again, curious about your thoughts on this

**Comment by timdorr** [comment]
We tell people to only put "serializable" data into the store. So, things that `JSON.stringify` can work with. But, despite our warnings, people still put all manner of silly things in there. 

As a practicality, I don't think we can restrict to objects and scalars.

I also agree that while `unknown` is probably more proper, it's a giant headache for users who already complain enough about Redux's boilerplate. 

> 🎉 New in Redux 5.0! More boilerplate!

**Comment by markerikson** [comment]
Yep.  Please don't constrain the state type.

**Comment by cellog** [comment]
note that `DeepPartial` is no longer used. #3485 removed its usage, but did not remove it from `index.d.ts`. I think we should delete it since it serves no useful purpose, and note that as a breaking change in case anyone was importing it. It's a 1 liner, they can just copy/paste to use it:

```ts
export type DeepPartial<T> = {
  [K in keyof T]?: T[K] extends object ? DeepPartial<T[K]> : T[K]
}
```

Any objections?

**Comment by timdorr** [comment]
Yeah, I was waiting for the next major to drop it, since we export it and others may be using it. It can go for 5.0.

**Comment by jedmao** [comment]
Hey @timdorr and @markerikson, please let me know if I can help at all with this TypeScript conversion. I'm super comfortable with TypeScript and I don't want you guys to struggle. I also realize you're using this as a sort of learning opportunity, so best of luck!

I know you already know the TS basics, so I want to recommend that the most important features of TS you need to learn are [generics](http://www.typescriptlang.org/docs/handbook/generics.html) and [conditional types](https://www.typescriptlang.org/docs/handbook/advanced-types.html#conditional-types). Master these concepts and you will be a TypeScript god. :godmode: 

Fortunately, the [TS documentation](http://www.typescriptlang.org/docs/home.html) is super good.

**Comment by timdorr** [comment]
And we are now converted to TS! 💃

**Comment by markerikson** [comment]
This effort seems to have stalled a bit.

Per discussion with Tim, it sounds like the main conversion work is done, and now it's a question of what improvements can be made to the TS types (including potentially breaking changes).

What potential types improvements are we thinking of?

If the `compose` changes from #3568 are really an improvement, I'm fine with somebody resurrecting those as a new PR (as long as the discussion stays focused on just the types aspects).

I know there's also some discussion about `ActionTypes.INIT` in #3580 .

Anything else?

**Comment by cellog** [comment]
I don't think #3568 is a minefield worth navigating unless someone runs into a bug with the existing types.

I agree fully that #3580 needs to be fixed.

I would encourage doing a release after that and then fixing with point releases, it's the only way to know how the changes affect people :)

**Comment by timdorr** [comment]
Does anyone know what `test/typescript/` is for now? It doesn't appear to be used in any tests or linter runs. It's also chock full of errors. Is it left over from the conversion and didn't get cleaned up? Should we try and fix the errors?

**Comment by cellog** [comment]
yeah.. you removed it :)

https://github.com/reduxjs/redux/commit/4a8eca868217fd557ff83637c713404b7c8ff61f

I strongly suggest reverting this commit. `typings-tester` allows verifying that types actually fail when they should, something that is impossible otherwise.

**Comment by cellog** [comment]
I feel I was a bit hasty in tossing off a quick reply. Let me elaborate.

I used the tests in `tests/typescript` while porting typescript over. They caught many breakages that I would have otherwise missed when making "minor" changes to the types. I was not familiar with `typings-tester` prior to working on this, and I became a big fan because of this. I think that this is the only way to make a fix that this bug requires while ensuring nothing else breaks.

There are, of course, probably other ways to do this, but since the infrastructure is already here, I recommend using it.

**Comment by cellog** [comment]
one last comment - the only problem in `typescript.spec.ts` was the comment, which read:

```ts
import { checkDirectory } from 'typings-tester'

describe('TypeScript definitions', function() {
  it('should compile against index.d.ts', () => {
    checkDirectory(__dirname + '/typescript')
  })
})
```

and should be something like:

```ts
import { checkDirectory } from 'typings-tester'

describe('TypeScript definitions', function() {
  it('should verify types fail when appropriate as well as working with normal usage', () => {
    checkDirectory(__dirname + '/typescript')
  })
})
```

**Comment by nickmccurdy** [comment]
Use [dtslint](https://github.com/microsoft/dtslint) instead, it's officially supported by Microsoft.

**Comment by cellog** [comment]
nice! following the link at dtslint which says "if you only need ExpectType and ExpectError" we get https://github.com/SamVerschueren/tsd

What do you think about using that @nickmccurdy ?

**Comment by nickmccurdy** [comment]
Not sure, I haven’t heard of it, but sounds worth trying

**Comment by jedmao** [comment]
Use neither.

> If you are writing the library in TypeScript, don't use dtslint. Use --declaration to have type definitions generated for you.

https://github.com/microsoft/dtslint#setup

The basic point is that you don't have to "test the types" if the types were generated, as you can trust they are accurate. This simplifies libraries that are written in TypeScript.

Please, don't reintroduce these types of tests.

**Comment by nickmccurdy** [comment]
Ah, I missed that. Have we made sure those test cases are tested in our new TypeScript tests?

**Comment by markerikson** [comment]
@jedmao : I'd disagree with that.  We're using some additional types tests over in Redux Starter Kit to make sure the types actually behave the way we want, and they've found several important problems.  Example:

https://github.com/reduxjs/redux-starter-kit/pull/209

**Comment by cellog** [comment]
Again, the tests check for things that are not otherwise possible to test for. Simply assuming they work without testing that they actually fail on invalid cases is risky for a library

**Comment by nickmccurdy** [comment]
Are you suggesting that we should assert both type errors and runtime errors using dynamic tests?

**Comment by nickmccurdy** [comment]
@cellog Are you suggesting that we should assert both type errors and runtime errors using dynamic tests?

There are some scenarios where types may be incorrect even if the runtime behavior is correct. I think @markerikson has a good point that we should still keep some sort of type tests for invalid cases.

**Comment by jedmao** [comment]
That's a good example, but I wouldn't apply it wholesale across the entire library. Just where it's relevant.

Looks like [tsd](https://github.com/SamVerschueren/tsd) is the better choice, as it's limited in scope and doesn't use the antiquated TSLint.
